warehouse-loader/warehouse-loader.py

```diff
@@ -0,0 +1,249 @@
+            return False
+        else:
+            raise ClientError
+    else:
+        return True
+
+
+def get_date_from_key(key, prefix):
+    """ Extract date from an object key from the bucket's directory pattern,
+    for a given prefix
+    
+    :param key: the object key in queustion
+    :type key: string 
+    :param prefix: the prefix to use, e.g. `raw/`, including the 
+    :type prefix: string
+    :return: the extracted date if found
+    :rtype: string or None
+    """
+    date_match = re.match(rf"^{prefix}(?P<date>[\d-]*)/.*", key)
+    if date_match:
+        return date_match.group("date")
+
+
+def patient_in_training_set(patient_id, training_percent=TRAINING_PERCENTAGE):
+    """ Separating patient ID's into training and validation sets, check
+    which one this ID should fall into.
+
+    It uses a hashing (sha512) to get pseudo-randomisation based on ID,
+    and do the cut-off with a set percentage.
+
+    :param patient_id: the candidate patient ID
+    :type patient_id: string 
+    :param training_percent: the percentage of patience to assign to the training set (defaults to the global TRAINING_PERCENTAGE)
+    :type training_percent: int
+    :return: True if the patient ID should fall into the training set
+    :rtype: boolean
+    """
+    return (
+        int(hashlib.sha512(patient_id.encode("utf-8")).hexdigest(), 16) % 100
+        < training_percent
+    )
+
+
+def extract_raw_folders():
+    """ Extractor: get all date folders within the `raw/` data drop
+
+    :return: subfolders within the `raw/` prefix (yield)
+    :rtype: string
+    """
+    result = s3_client.list_objects(
+        Bucket=BUCKET_NAME, Prefix=RAW_PREFIX, Delimiter="/"
+    )
+    for subfolder in result.get("CommonPrefixes"):
+        yield subfolder.get("Prefix")
+
+
+def extract_raw_files_from_folder(folder):
+    """ Extract files from a givem date folder in the data dump
+    
+    :param folder: the folder to process
+    :type key: string 
+    :return: each object (yield)
+    :rtype: boto3.resource('s3').ObjectSummary
+    """
+    for obj in bucket.objects.filter(Prefix=folder):
+        yield obj
+
+
+def process_image(obj):
+    """ Processing images from the raw dump
+
+    Takes a single image, downloads it into temporary storage
+    and extracts its metadata.
+
+    The metadata is then uploaded here, except if the file already exists.
+
+    If the image file already exists at the correct location, it's not passed
+    on to the next step.
+    
+    :param obj: the object in queustion
+    :type key: boto3.resource('s3').ObjectSummary
+    :return: the original object, and a new key where it should be copied within the bucket
+    :rtype: (boto3.resource('s3').ObjectSummary, string)
+    """
+    result = (None, None)
+    tmp = tempfile.NamedTemporaryFile(suffix=".dcm", delete=False)
+    with open(tmp.name, "wb") as fd:
+        obj.Object().download_file(tmp.name)
+    with open(tmp.name, "rb") as fd:
```

Comment:
As mentioned, I think the file could be read directly into memory. See e.g.:

https://stackoverflow.com/questions/44043036/how-to-read-image-file-from-s3-bucket-directly-into-memory

Reply:
This code is before your feedback last weekend, and haven't added those changes yet. Thanks for the direct link for extra context. Though that you linked is not much different, just fewer lines of code, but it still uses `tempfile.NamedTemporaryFile()`, so both mine and SO both keeps things in memory (on tmpfs).  I have actually tried something very similar, but then `pydicom` was unhappy, hence the extra reopen, but revisiting this next.

Reply:
The second answer puts it directly into a StringIO object. The third answer uses BytesIO, which is probably even more appropriate?

Reply:
Yeah, that's fair, I didn't check that. And indeed we talked about BytesIO, thanks for the patience! :)